### ISSUE TITLE
Bug in animated implementation

### ISSUE THREAD

**Issue by jfrolich** [reason-react-native/reason-react-native#620]
`createAnimatedComponent` needs to be hoisted out of the render function for it to work correctly. (It seems to work correctly but it has strange side-effects like scroll position resetting).

I did not find a way to do this without type errors. So couldn't publish a pull-request yet. Help welcome!

https://github.com/reason-react-native/reason-react-native/blob/master/src/apis/Animated.re#L301

**Comment by jfrolich** [comment]
```reason
module AnimatedSectionList = {
  include SectionList;

  let animatedSectionList = Obj.magic(Animated.createAnimatedComponent(SectionList.make));
  let make = props =>
    React.createElementVariadic(animatedSectionList, props, [||]);
};
```
This compiles. But there might be a better option :D

**Comment by sgny** [comment]
This is a bit curious. I didn't test with `SectionList` (because I don't use that component) but with `FlatList` and everything was alright, no artifacts at all. Now neither `Animated.FlatList` nor `Animated.SectionList` can be found.

However, if you simply copy-paste both implementation into the module, both compile and the `JS` output is identical (except one subsitution).

with `Obj.magic`
```
var animatedSectionList = ReactNative.Animated.createAnimatedComponent(ReactNative.SectionList);

function make(props) {
  return React.createElement(animatedSectionList, props);
}
```

without
```
function make(props) {
  return React.createElement(ReactNative.Animated.createAnimatedComponent(ReactNative.SectionList), props);
}
```

Could you provide a sample code using an animated `SectionList` so I can do further testing?

Also, what's your `bs-platform` version? I have moved onto version 6.2 recently, but I was using a stable v5 release when I added the implementation.

**Comment by sgny** [comment]
It gets weirder, but at least what I'm seeing has nothing to do with using pre-production `bs-platform`.

I use `reason-react-native` and `react-native` from git. Recently, I've been using [`4d6e32b`]
(https://github.com/reason-react-native/reason-react-native/tree/4d6e32bd4210488afc2e57f9c7d6c490b62b5605), which is only two commits behind. Neither of those commits touch code.

Yet, I don't have `SectionList` or `FlatList` in `Animated.re` which is why they could not be found by `bsb` (no surprise there).

I saw another report of incorrect version getting installed by `npm` (with version 6.11.3, I'm on 6.12). Perhaps it's time I switched to `yarn` ☹️.

**Comment by jfrolich** [comment]
> This is a bit curious. I didn't test with `SectionList` (because I don't use that component) but with `FlatList` and everything was alright, no artifacts at all. Now neither `Animated.FlatList` nor `Animated.SectionList` can be found.
> 
> However, if you simply copy-paste either implementation into a module, both compile and the `JS` output is identical (except one subsitution).
> 
> with `Obj.magic`
> 
> ```
> var animatedSectionList = ReactNative.Animated.createAnimatedComponent(ReactNative.SectionList);
> 
> function make(props) {
>   return React.createElement(animatedSectionList, props);
> }
> ```
> 
> without
> 
> ```
> function make(props) {
>   return React.createElement(ReactNative.Animated.createAnimatedComponent(ReactNative.SectionList), props);
> }
> ```
Yes that is exactly what needs to happen, createAnimatedComponent should not be in render, because it is a higher order component yielding a class with internal state. If it's hoisted it is always the same class, and it will not be recreated on every render.

See implementation here: https://github.com/facebook/react-native/blob/91f139b94118fe8db29728ea8ad855fc4a13f743/Libraries/Animated/src/createAnimatedComponent.js.

The `Obj.magic` is the only way how I could get it to compile, but there should be a better way.

You can reproduce this with a sectionList where you load some new data (it will re-render the list and reset the scroll position while doing that).

**Comment by jfrolich** [comment]
> I use `reason-react-native` and `react-native` from git. Recently, I've been using [`4d6e32b`](https://github.com/reason-react-native/reason-react-native/tree/4d6e32bd4210488afc2e57f9c7d6c490b62b5605), which is only two commits behind. Neither of those commits touch code.

Yeah, I think SectionList in animated is only in master yet.

**Comment by sgny** [comment]
Well, I always specify a commit and not `master`. In this case, the commit should have had all these. That's why I was surprised, it's most likely a bug.

I probably won't be able to test until much later today, but to clarify, for you, your version compiles but mine does not, is that the case?

And with your version, do you get artifacts or does it work as expected?

**Comment by jfrolich** [comment]
Actually found a shorter version that works:

```reason
module AnimatedSectionList = {
  include SectionList;
  let make = Obj.magic(Animated.createAnimatedComponent(SectionList.make));
};
```

**Comment by jfrolich** [comment]
This codes works If you have `SectionList` (opened ReactNative)

**Comment by sgny** [comment]
If `Obj.magic` is omitted, the compiler complains of types that cannot be generalized, so that's neat way to get around that. Kudos.

> This codes works

Do you mean the issue with scroll position getting reset no longer arises?

**Comment by jfrolich** [comment]
> Do you mean the issue with scroll position getting reset no longer arises?

Yes the HoC is not applied in render. This results in the same code as the original implementation in the JavaScript version of Animated.

**Comment by sgny** [comment]
To be honest, I don't see any reason why there should be such a difference, because looking at the JS output, there is only an additional `createElement` layer in the older implementation. However, I had already noted that in the [PR](https://github.com/reason-react-native/reason-react-native/pull/600) as an issue to be resolved. I did not think to resolve the type error by typecasting; it's nice that you have. Perhaps the props are passed improperly due to that layer.

Then, even if there were not glitches resolved by this, removing the superfluous `createElement` would be a worthy improvement.

**Comment by jfrolich** [comment]
> To be honest, I don't see any reason why there should be such a difference, because looking at the JS output, there is only an additional `createElement` layer in the older implementation. However, I had already noted that in the [PR](https://github.com/reason-react-native/reason-react-native/pull/600) as an issue to be resolved. I did not think to resolve the type error by typecasting; it's nice that you have. Perhaps the props are passed improperly due to that layer.
> 
> Then, even if there were not glitches resolved by this, removing the superfluous `createElement` would be a worthy improvement.

The minor difference you see is that the higher order component `createAnimatedComponent` is not created in the render function. I'm not sure how familiar you are with React, but that is a huge difference because in render the component is re-created on every render (reseting all internal state and wiping out, and rebuilding the entire tree because it yields a different component), and in my implementation the animated component is only created exactly once.

**Comment by sgny** [comment]
Actually the confusion is mostly because I have limited experience with JavaScript. ☹️ I needed to be able to create a frontend, so I have started working with PureScript and moved over to Reason.

With your prompting, now the difference is clear.

#### output 1:
```
function make(props) {
  return React.createElement(ReactNative.Animated.createAnimatedComponent(ReactNative.SectionList), props);
}
```
is a function and accordingly only executed when `make` is called (so the Animated component would be created then) and not when declared

#### output 2:
```
var make = ReactNative.Animated.createAnimatedComponent(ReactNative.SectionList);
```
is executed when declared.

This has been quite informative. It appears, I will have to look further into JS at some point.

Now that we have established that the implementation has to change, would you like to create a PR or should I take care of that?

**Comment by jfrolich** [comment]
I can make a PR 👍

**Comment by sgny** [comment]
Implemented in #623.
